### DISCUSSION TITLE
Increase of a variable residual and convergence issue

### DISCUSSION THREAD

**Discussion by sinaomrani96** [idaholab/moose#25072] [Q&A General]
Hi,
Hope you are well.

I am using PorousFlow module of MOOSE but my question is a bit more general and related to solving the equations. I have turned on the show_var_residual_norms to see the residual for each variable. The thing that I do not understand is that somewhere in the solve, a variable residual increases and causes convergence problem (please see below the residual for tracer variable). I was wondering what is the reason and how I can solve it. Thanks.

Time Step 3, time = 3, dt = 1
 0 Nonlinear |R| = 2.211444e-06
    |residual|_2 of individual variables:
                     pwater: 9.11345e-09
                     satg:   9.16047e-22
                     tracer: 2.21143e-06

 1 Nonlinear |R| = 8.379444e-06
    |residual|_2 of individual variables:
                     pwater: 5.11939e-07
                     satg:   7.53075e-22
                     tracer: 8.36379e-06

 2 Nonlinear |R| = 6.201440e-05
    |residual|_2 of individual variables:
                     pwater: 3.9068e-06
                     satg:   5.89635e-22
                     tracer: 6.18912e-05

 3 Nonlinear |R| = 6.116279e-05
    |residual|_2 of individual variables:
                     pwater: 3.41119e-06
                     satg:   4.76689e-22
                     tracer: 6.10676e-05

Best regards,
Sina

**Comment by lindsayad** [comment]
I would start by reading https://mooseframework.inl.gov/moose/application_usage/failed_solves.html if you haven't already. Seeing a variable residual increase during an iteration of a Newton solve is not unusual, but if you're unable to run your simulation satisfactorily then there may be issues. Are you using Newton or PJFNK as your `solve_type`? How are your linear solves?

**Reply by sinaomrani96** [reply]
Hi Alex,

Thanks for sharing the link. I am using Newton as my solver. The linear solve is also slow. I have tested various preconditioning for my problem and only the following setting gave me better results. I have been testing different nl_abs_tol and l_abs_tol to find the one suitable for my problem.

  [precond]
    type = SMP
    full = true
    petsc_options = '-snes_linesearch_monitor'
    petsc_options_iname = '-snes_type -pc_factor_shift_type -snes_linesearch_type'
    petsc_options_value = 'vinewtonssls nonzero l2'
  []

To be honest, I do not know so much about various preconditioning settings, and how I can choose one based on my problem. I also appreciate if you have any suggestion in this regard. Thanks.

**Reply by lindsayad** [reply]
What physics are you solving? You have to be solving very special problems in order to use `vinewtonssls`

**Reply by sinaomrani96** [reply]
I am using PorousFlow and my problem is a multi-component fluid flow in porous media. However, the viscosity is not constant (it is a non-newtonian fluid). As I said, I do not know about the preconditioning and found this one in a tensor mechanics example I think.

**Reply by lindsayad** [reply]
@cpgr and @WilkAndy can advise you on good preconditioning settings. For debugging your convergence and ensuring the linear solves aren't the problem, I would suggest changing to
```
[precond]
  solve_type = NEWTON
  type = SMP
  full = true
  petsc_options = '-snes_linesearch_monitor'
  petsc_options_iname = '-pc_type -pc_factor_shift_type -snes_linesearch_type'
  petsc_options_value = 'lu NONZERO basic'
[]
```

**Reply by GiudGiud** [reply]
iirc there is a porous flow preconditioning page in the porous flow documentation you should consult

**Reply by cpgr** [reply]
@WilkAndy has provided some good suggestions here:

https://mooseframework.inl.gov/modules/porous_flow/solvers.html
https://mooseframework.inl.gov/modules/porous_flow/nonlinear_convergence_problems.html
and also
https://mooseframework.inl.gov/modules/porous_flow/convergence.html
